[Zscaler] Add URLs to Retaining Parent Category

samuelFain opened the pull request:
<!-- REMINDER: THIS IS A PUBLIC REPO DO NOT POST HERE SECRETS/SENSITIVE DATA -->
## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: [XSUP-29447](https://jira-dc.paloaltonetworks.com/browse/XSUP-29447)

## Description
This PR introduce the following changes:
- Add "retaining-parent-category-url/ip" arguments for the add commands (zscaler-category-add-url, zscaler-category-add-ip).
- Add "retaining-parent-category-url/ip" arguments for the remove commands (zscaler-category-remove-url, zscaler-category-remove-ip).
- For the mentioned commands, the `url` argument is no longer mandatory. Instead, there are now two available arguments (`url` / `retaining_parent_category`), where at least one must be used, otherwise an error message will be printed.
- Linting & Formatting.

## Must have
- [x] Tests
- [x] Documentation

JasBeilin commented on Packs/Zscaler/Integrations/Zscaler/Zscaler.py:
```diff
@@ -607,15 +608,20 @@ def lookup_request(ioc, multiple=True):
         category_data["urls"] = all_urls
+        retaining_parent_category_url_list = argToList(retaining_parent_category_url)
+        if not (url_list, retaining_parent_category_url_list):
+            return_error('Either url argument or retaining_parent_category_url argument must be provided.')
```
```suggestion
            return_error('Please provide retaining_parent_category_url or url argument.')
```

JasBeilin commented on Packs/Zscaler/Integrations/Zscaler/Zscaler.py:
```diff
@@ -644,7 +650,10 @@ def category_add_url(category_id, url):
-def category_add_ip(category_id, ip):
+def category_add_ip(category_id, ip, retaining_parent_category_ip):
+    if not (ip, retaining_parent_category_ip):
+        return_error('Either ip argument or retaining_parent_category_ip argument must be provided.')
```
```suggestion
        return_error('Please provide ip or retaining_parent_category_ip argument.')
```

JasBeilin commented on Packs/Zscaler/Integrations/Zscaler/Zscaler.py:
```diff
@@ -657,7 +666,10 @@ def category_add_ip(category_id, ip):
-        response = category_ioc_update(category_data)
+        retaining_parent_category_ip_list = argToList(retaining_parent_category_ip)
+        if not (ip_list, retaining_parent_category_ip_list):
+            return_error('Either ip_list argument or retaining_parent_category_ip_list argument must be provided.')
```
```suggestion
            return_error('Please provide ip_list argument or retaining_parent_category_ip_list argument.')
```

JasBeilin commented on Packs/Zscaler/Integrations/Zscaler/Zscaler.py:
```diff
@@ -687,19 +699,35 @@ def category_add_ip(category_id, ip):
-def category_remove_url(category_id, url):
+def category_remove_url(category_id, url, retaining_parent_category_url):
+    if not (url, retaining_parent_category_url):
+        return_error('Either url argument or retaining_parent_category_url argument must be provided.')
```
```suggestion
        return_error('Please provide  url or retaining_parent_category_url argument.')
```

JasBeilin reviewed (commented):
Looks good, please change the errors to a clear action.
Also, please add test for the remove as well

samuelFain commented on Packs/Zscaler/Integrations/Zscaler/Zscaler.py:
```diff
@@ -281,7 +282,7 @@ def unblacklist_ip(ip):
 
 
-def whitelist_url(url):
+def whitelist_url(url):  # pragma: no cover
```
Removed, I already added relevant UTs to changed functions (this is not one of them).

samuelFain commented on Packs/Zscaler/Integrations/Zscaler/Zscaler.py:
```diff
@@ -298,7 +299,7 @@ def whitelist_url(url):
 
 
-def unwhitelist_url(url):
+def unwhitelist_url(url):  # pragma: no cover
```
Removed, I already added relevant UTs to changed functions (this is not one of them).

samuelFain commented on Packs/Zscaler/Integrations/Zscaler/Zscaler_test.py:
```diff
@@ -23,7 +23,7 @@ class ObjectMocker(dict):
 
 
 def run_command_test(command_func, args, response_path, expected_result_path, mocker, result_validator=None):
-    with open(response_path, 'r') as response_f:
```
This change is a part of automated ruff linter changes.

samuelFain commented on Packs/Zscaler/Integrations/Zscaler/Zscaler_test.py:
```diff
@@ -36,7 +36,7 @@ def run_command_test(command_func, args, response_path, expected_result_path, mo
     if result_validator:
         assert result_validator(res)
     else:
-        with open(expected_result_path, 'r') as ex_f:
```
This change is a part of automated ruff linter changes.

samuelFain commented:
> Looks good, please change the errors to a clear action. Also, please add test for the remove as well

I've addressed your comments, please re-review 🙂.
UTs for modified `remove` commands already exist:
- test_category_remove_ip
- test_category_remove_url
